sdk/clientcore/core/src/main/java/io/clientcore/core/util/union/Union.java

```diff
@@ -0,0 +1,284 @@
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsCollectionType -->
+ *
+ * <p><strong>Consume the value of the Union if it is of the expected type</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * switch &#40;union.getValue&#40;&#41;&#41; &#123;
+ *     case String s -&gt; System.out.println&#40;&quot;String value: &quot; + s&#41;;
+ *     case Integer i -&gt; System.out.println&#40;&quot;Integer value: &quot; + i&#41;;
+ *     default -&gt; throw new IllegalArgumentException&#40;
+ *         &quot;Unknown type: &quot; + union.getType&#40;&#41;.getTypeName&#40;&#41;&#41;;
+ * &#125;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ *
+ * or
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;String value: &quot; + v&#41;, String.class&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;Integer value: &quot; + v&#41;, Integer.class&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ *
+ */
+public final class Union {
+    private static final ClientLogger LOGGER = new ClientLogger(Union.class);
+
+    private final List<Type> types;
+    private Object value;
+    private Type currentType;
+
+    private Union(Type... types) {
+        this.types = Arrays.asList(types);
+    }
+
+    /**
+     * Creates a new instance of {@link Union} with the provided types.
+     *
+     * @param types The types of the union.
+     * @return A new instance of {@link Union}.
+     */
+    public static Union ofTypes(Type... types) {
+        return new Union(types);
+    }
+
+    /**
+     * Sets the value of the union.
+     *
+     * @param value The value of the union.
+     * @throws IllegalArgumentException If the value is not of one of the types in the union.
+     */
+    @SuppressWarnings("unchecked")
+    public void setValue(Object value) {
+        for (Type type : types) {
+            if (isInstanceOfType(value, type) || isPrimitiveTypeMatch(value, type)) {
+                this.value = value;
+                this.currentType = type;
+                return;
+            }
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Invalid type: " + value.getClass().getName()));
+    }
+
+    /**
+     * Gets the type of the value.
+     *
+     * @return The type of the value.
+     */
+    public Type getType() {
+        return currentType;
+    }
+
+    /**
+     * Gets the types of the union.
+     *
+     * @return The types of the union.
+     */
+    public List<Type> getTypes() {
```

Comment:
Should we call this getSupportedTypes or getVariantTypes or something?

Reply:
I think a better name for this is a good idea. Another suggestion - `getUnionTypes`

Reply:
`union.getSupportedTypes()` sounds better to me in my opinion than `union.getUnionTypes()`. Will stay with getSupportedTypes for now and will update it if found better suitable name.